### PR TITLE
Do not parse CSS on missed plugins or parser/stringifier option

### LINKED ISSUE #1161: Your current PostCSS version is 5.2.18, but postcss-modules-scope uses 6.0.21. Perhaps this is the source of the error below.
**Do you want to request a *feature* or report a *bug*?**
**BUG**

**What is the current behavior?**
When I try to start my webpack bundle compilation I even gets the strange error about incomportability of `PostCSS` and `postcss-modules-scope` versions. What is the interesting that is I had installed the lastest version of PostCSS (6.0.23), but the systam says that I got only PostCSS version is 5.2.18. 

```
Your current PostCSS version is 5.2.18, but postcss-modules-scope uses 6.0.21. Perhaps this is the source of the error below.
ERROR in webpack_test/react-apps/shared/components/LoadingIndicator/styles.cssmodule
  app:build:webpack-compiler Module build failed (from /webpack_test/react-apps/node_modules/mini-css-extract-plugin/dist/loader.js):
  app:build:webpack-compiler ModuleBuildError: Module build failed (from /webpack_test/react-apps/node_modules/css-loader/index.js):
  app:build:webpack-compiler Error: composition is only allowed when selector is single :local class name not in ".dot1", ".dot1" is weird
  app:build:webpack-compiler     at \webpack_test\react-apps\node_modules\postcss-modules-scope\lib\index.js:26:13
```
**If the current behavior is a bug, please provide the steps to reproduce.**
It's start happen after I had updated my webpack from 3.8.1 to 4.12.1 version. That's all.

**What is the expected behavior?**
Webpack compilation should proceed normally. Without any crashes on css step.

My webpack file:
```
const webpack = require('webpack')
const HtmlWebpackPlugin = require('html-webpack-plugin')
const MiniCssExtractPlugin = require('mini-css-extract-plugin')
const UglifyJsPlugin = require('uglifyjs-webpack-plugin')
const debug = require('debug')('app:webpack:config')

// ------------------------------------
// Rules
// ------------------------------------
const rules = []

// JavaScript / JSON
rules.push(
  {
    test: /\.(js|jsx)$/,
    exclude: /node_modules/,
    loader: 'babel-loader',
    query: {
      cacheDirectory: true,
      plugins: [
        'babel-plugin-transform-class-properties',
        'babel-plugin-syntax-dynamic-import',
        [
          'babel-plugin-transform-runtime',
          {
            helpers: true,
            polyfill: true,
            regenerator: true
          }
        ],
        [
          'babel-plugin-transform-object-rest-spread',
          {
            useBuiltIns: false
          }
        ]
      ],
      presets: [
        'babel-preset-react',
        [
          'babel-preset-env',
          {
            modules: false,
            targets: {
              ie9: true
            },
            uglify: true
          }
        ]
      ]
    }
  },
  {
    test: /\.json$/,
    loader: 'json-loader'
  }
)

// ------------------------------------
// Style Loaders
// ------------------------------------

rules.push(
  {
    test: /\.(s?)css$/,
    use: [
      MiniCssExtractPlugin.loader,
      { loader: 'css-loader', options: { sourceMap: false } },
      { loader: 'postcss-loader', options: { sourceMap: false } },
      { loader: 'sass-loader', options: { sourceMap: false } }
    ]
  },
  {
    test: /\.cssmodule$/,
    use: [
      MiniCssExtractPlugin.loader,
      {
        loader: 'css-loader?modules&importLoaders=1&localIdentName=[local]___[hash:base64:5]',
        options: { sourceMap: false }
      },
      { loader: 'postcss-loader', options: { sourceMap: false } },
      { loader: 'sass-loader', options: { sourceMap: false } }
    ]
  }
)

// File loaders
/* eslint-disable */
rules.push(
  {
    test: /\.woff(\?.*)?$/,
    loader: 'url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff'
  },
  {
    test: /\.woff2(\?.*)?$/,
    loader: 'url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/font-woff2'
  },
  {
    test: /\.otf(\?.*)?$/,
    loader: 'file-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=font/opentype'
  },
  {
    test: /\.ttf(\?.*)?$/,
    loader: 'url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=application/octet-stream'
  },
  {
    test: /\.eot(\?.*)?$/,
    loader: 'file-loader?prefix=fonts/&name=[path][name].[ext]'
  },
  {
    test: /\.svg(\?.*)?$/,
    loader: 'url-loader?prefix=fonts/&name=[path][name].[ext]&limit=10000&mimetype=image/svg+xml'
  },
  { test: /\.(png|jpg)$/, loader: 'url-loader?limit=8192' }
)

const createConfig = options => {
  const paths = options.utils_paths
  const __DEV__ = options.globals.__DEV__
  const __PROD__ = options.globals.__PROD__
  // TODO: sourceMap is not read from postcss.config.js because of a bug
  // remove it once postcss-loader is updated

  debug('Creating configuration.')
  const webpackConfig = {
    mode: __DEV__ ? 'development' : 'production',
    entry: paths.client('main.js'),
    name: 'client',
    target: 'web',
    devtool: options.compiler_devtool,
    optimization: {
      minimizer: [
        new UglifyJsPlugin({
          uglifyOptions: {
            compress: {
              unused: true,
              dead_code: true,
              warnings: false
            }
          },
          sourceMap: false
        })
      ],
      splitChunks: {
        chunks: 'all'
      }
    },
    performance: {
      hints: false
    },
    resolve: {
      modules: [paths.client(), 'node_modules'],
      extensions: ['.js', '.jsx', '.json']
    },
    module: {}
  }

  // ------------------------------------
  // Entry Points
  // ------------------------------------
  const APP_ENTRY = paths.client('main.js')

  webpackConfig.entry = {
    app: __DEV__
      ? [APP_ENTRY].concat(`webpack-hot-middleware/client?path=${options.compiler_public_path}__webpack_hmr`)
      : [APP_ENTRY]
  }

  // ------------------------------------
  // Bundle externals
  // ------------------------------------
  webpackConfig.externals = {
    react: 'React',
    'react-dom': 'ReactDOM'
  }

  // ------------------------------------
  // Bundle Output
  // ------------------------------------
  webpackConfig.output = {
    filename: `[name].[${options.compiler_hash_type}].js`,
    path: paths.dist(),
    publicPath: options.compiler_public_path
  }

  // ------------------------------------
  // Plugins
  // ------------------------------------
  webpackConfig.plugins = [
    new webpack.DefinePlugin(options.globals),
    new MiniCssExtractPlugin({
      filename: '[name].[hash].css',
      chunkFilename: '[name].[hash].css'
    })
  ]

  if (__DEV__) {
    debug('Enable plugins for live development (HMR, NoErrors).')
    webpackConfig.plugins.push(
      new HtmlWebpackPlugin({
        entry: {
          app: paths.client('index.html')
        },
        template: paths.client('index.html'),
        filename: 'index.html',
        inject: 'body',
        minify: {
          collapseWhitespace: true
        }
      }),
      new webpack.HotModuleReplacementPlugin(),
      new webpack.NoEmitOnErrorsPlugin()
    )
  } else if (__PROD__) {
    debug('Enable plugins for production')
    webpackConfig.plugins.push(
      new HtmlWebpackPlugin({
        template: paths.client(options.template),
        filename: options.template,
        minify: false,
        inject: false
      })
    )

    webpackConfig.plugins.push(
      new webpack.ProvidePlugin({
        fetch: 'exports-loader?self.fetch!whatwg-fetch'
      })
    )
  }

  webpackConfig.module.rules = rules

  return webpackConfig
}

module.exports.rules = rules

module.exports = createConfig

```

My package.json file:
```
{
  "name": "react-apps",
  "version": "0.0.1",
  "description": "",
  "private": true,
  "workspaces": [
    "apps/*",
    "shared"
  ],
  "pre-commit": "lint:staged",
  "lint-staged": {
    "*.{js,jsx}": [
      "prettier --single-quote --no-semi --print-width 120 --write",
      "git add"
    ]
  },
  "scripts": {
    "deployall": "lerna run deploy --concurrency 1",
    "lint:staged": "lint-staged",
    "storybook": "start-storybook -p 6006",
    "test": "jest",
    "test:dev": "npm run test -- --watch",
    "build-storybook": "build-storybook"
  },
  "jest": {
    "verbose": true,
    "transform": {
      "^.+\\.js$": "babel-jest"
    },
    "moduleNameMapper": {
      "^.+\\.(css|scss|cssmodule)$": "identity-obj-proxy"
    },
    "globals": {
      "NODE_ENV": "test"
    },
    "moduleFileExtensions": [
      "js"
    ],
    "transformIgnorePatterns": [
      ".*(node_modules)(?!.*.*).*$"
    ]
  },
  "author": "",
  "devDependencies": {
    "@storybook/addon-actions": "^3.2.17",
    "@storybook/addon-backgrounds": "^3.3.10",
    "@storybook/addon-centered": "^3.3.10",
    "@storybook/addon-knobs": "^3.3.10",
    "@storybook/addon-links": "^3.2.17",
    "@storybook/cli": "^3.4.2",
    "@storybook/react": "^3.2.17",
    "babel-eslint": "^8.2.3",
    "babel-plugin-transform-object-assign": "^6.22.0",
    "babel-polyfill": "^6.26.0",
    "babel-preset-env": "^1.7.0",
    "babel-preset-react": "^6.24.1",
    "babel-template": "^6.26.0",
    "babel-types": "^6.26.0",
    "css-loader": "^0.28.11",
    "eslint": "^4.19.1",
    "eslint-config-standard": "^11.0.0",
    "eslint-config-standard-react": "^6.0.0",
    "eslint-plugin-babel": "^5.0.0",
    "eslint-plugin-promise": "^3.4.0",
    "eslint-plugin-react": "^7.7.0",
    "eslint-plugin-standard": "^3.0.1",
    "extract-text-webpack-plugin": "^4.0.0-beta.0",
    "fetch-mock": "^6.0.0",
    "glob-loader": "^0.3.0",
    "html-webpack-plugin": "^3.2.0",
    "lint-staged": "^7.0.4",
    "mini-css-extract-plugin": "^0.4.0",
    "postcss": "^6.0.23",
    "postcss-loader": "^2.1.5",
    "postcss-modules-scope": "^1.1.0",
    "pre-commit": "^1.2.2",
    "prettier": "^1.10.2",
    "prop-types": "^15.6.0",
    "redux-mock-store": "^1.5.1",
    "ssh2-sftp-client": "^2.1.1",
    "uglifyjs-webpack-plugin": "^1.2.6",
    "webpack": "^4.12.1",
    "webpack-dev-middleware": "^3.1.3",
    "webpack-hot-middleware": "^2.22.2",
    "webpack-middleware": "^1.5.1"
  },
  "dependencies": {
    "@types/jest": "^22.2.3",
    "eslint-plugin-import": "^2.11.0",
    "react": "^16.3.2",
    "react-dom": "^16.3.2"
  }
}

```

**Please mention other relevant information such as your webpack version, Node.js version and Operating System.**
Windows 10,
NodeJs 6.0.1
Yarn 1.7.0
Webpack  4.12.1
PostCSS 6.0.23
postcss-modules-scope 1.1.0
css-loader 2.1.5
@ai #Your current PostCSS version is 5.2.18, but postcss-modules-scope uses 6.0.21

### LINKED ISSUE #1651: Do not parse CSS on missed plugins or parser/stringifier option
Right now we are showing warning:

> You did not set any plugins, parser, or stringifier. Right now, PostCSS does nothing. Pick plugins for your case on https://www.postcss.parts/ and use them in postcss.config.js.

https://github.com/postcss/postcss/blob/main/lib/processor.js#L19

But many people complain that they can fix it since PostCSS is used wrongly somewhere inside build pipeline.

We can avoid warning if PostCSS will not parse CSS in this case. We can add something like `NoWorkResult`, where `NoWorkResult#css` key will be equal to input. On `NoWorkResult#root` read we still need to parse CSS. And we need to find the best way to deal with `NoWorkResult#map`.

### PR DESCRIPTION
**Pull request by bogdan0083**
This is my attempt to fix #1651 issue.

There are some quirks in my implementation but I hope it's not too bad :)

### CODE REVIEW

File: lib/no-work-result.d.ts [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+ * are accessed. See the example below for details.
+ * A `NoWork` instance is returned by `Processor#process` ONLY when no plugins defined.
+ *
+ * ```js
+ * const noWorkResult = postcss().process(css) // No plugins are defined.
+ *                                             // CSS is not parsed
+ * let root = noWorkResult.root // now css is parsed because we accessed the root
+ * ```
+ */
+export default class NoWorkResult implements PromiseLike<Result> {
```
Comment (ai) [inline lib/no-work-result.d.ts:19]: Can we do `export default NoWorkResult extends LazyResult` and avoid copying docs for methods?

File: lib/processor.js [hunk 2]
```diff
@@ -3,10 +3,11 @@
 let LazyResult = require('./lazy-result')
 let Document = require('./document')
 let Root = require('./root')
+let NoWorkResult = require('./no-work-result')
 
 class Processor {
   constructor(plugins = []) {
-    this.version = '8.3.11'
+    this.version = '8.3.12'
```
Comment (ai) [inline lib/processor.js:10]: I will bump version in separated release commit

File: test/processor.test.ts [hunk 3]
```diff
@@ -525,14 +526,50 @@ it('warns about missed from', async () => {
...
+})
+
+it('catches error with empty processor', async () => {
+  let noWorkResult = new Processor().process('a {}')
+
+  // @ts-ignore
+  noWorkResult.error = new CssSyntaxError('error')
+
+  noWorkResult.catch(err => {
+    // eslint-disable-next-line jest/no-conditional-expect
```
Comment (ai) [inline test/processor.test.ts:570]: I removed Jest ESLint plugin in recent commit

```suggestion
```

File: package.json [hunk 4]
```diff
@@ -19,6 +19,7 @@
     "./lib/fromJSON": "./lib/fromJSON.js",
     "./lib/input": "./lib/input.js",
     "./lib/lazy-result": "./lib/lazy-result.js",
+    "./lib/no-work": "./lib/no-work.js",
```
Comment (ai) [inline package.json:22]:
```suggestion
    "./lib/no-work-result": "./lib/no-work-result.js",
```

File: lib/map-generator.js [hunk 5]
```diff
@@ -67,12 +68,17 @@ class MapGenerator {
...
 
-    let node
-    for (let i = this.root.nodes.length - 1; i >= 0; i--) {
-      node = this.root.nodes[i]
-      if (node.type !== 'comment') continue
-      if (node.text.indexOf('# sourceMappingURL=') === 0) {
-        this.root.removeChild(i)
+    if (!this.root && typeof this.css === 'string') {
+      // @NOTE: not sure about this regex. Is it safe to use?
+      this.css = this.css.replace(/\/\*#[\S\s]*?\*\/$/gm, '')
```
Comment (ai) [inline lib/map-generator.js:73]: Let’s test by having 2 `/*# sourceMappingURL=… */` comments with a code between them.

File: lib/no-work-result.js [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+
+let MapGenerator = require('./map-generator')
+let stringify = require('./stringify')
+let warnOnce = require('./warn-once')
+let parse = require('./parse')
+const Result = require('./result')
+
+// @FIXME: Not sure why eslint is complaining about this. Temporary disable.
+// eslint-disable-next-line
+let postcss = {}
```
Comment (ai) [inline lib/no-work-result.js:11]: What is the error?

File: lib/no-work-result.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    let root
+
+    let str = stringify
+    this.result = new Result(this._processor, root, this._opts)
+    this.result.css = css
+
+    // @NOTE: this one is tricky. We set Result#root getter to return
+    // the NoWorkResult#root instead, to preserve laziness on Result#root access
+    let self = this
+    Object.defineProperty(this.result, 'root', {
```
Comment (ai) [inline lib/no-work-result.js:31]: What if we will return `this` (self) on `NoWork#result`?

File: test/processor.test.ts [hunk 8]
```diff
@@ -525,14 +526,49 @@ it('warns about missed from', async () => {
...
+  await processor.process('a{}')
+  expect(spy).toHaveBeenCalledTimes(1)
+  spy.mockRestore()
+})
+
+it('catches error with empty processor', async () => {
+  let noWorkResult = new Processor().process('a {}')
+
+  // @ts-ignore
+  noWorkResult.error = new CssSyntaxError('error')
```
Comment (ai) [inline test/processor.test.ts:567]: Let’s use invalid CSS and `.root` access instead of the hack to have an error

**Comment by ai** [comment]
@bogdan0083 I left comments in changes diff. Can you review them?

**Comment by bogdan0083** [comment]
> @bogdan0083 I left comments in changes diff. Can you review them?

What do you mean? Can you see my responses to your review?


Or do you want me to resolve all the changes?

File: lib/map-generator.js (hunk 5 above)
Comment (ai) [reply lib/map-generator.js:73]: Looks fine. Let’s remove comment.

File: lib/map-generator.js [hunk 9]
```diff
@@ -176,6 +182,34 @@ class MapGenerator {
     return [this.css, this.map]
   }
 
+  generateSimpleMap() {
+    this.map = new SourceMapGenerator({ file: this.outputFile() })
+    this.previousMaps = []
+
```
[the remark this answers is not in the document]
Comment (ai) [reply lib/map-generator.js:188]: Why not just:

```js
      this.map.addMapping({
        source,
        generated: { line: 0, column: 0 },
        original: { line: 0, column: 0 }
      })
```

You can use [this tool](https://sokra.github.io/source-map-visualization/) to check source map content.

File: lib/no-work-result.js (hunk 6 above)
Comment (ai) [reply lib/no-work-result.js:11]: But why do you need to assign `postcss` variable if you do not use it in this file?

File: test/no-work-result.test.ts [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+  let result = new NoWorkResult(processor, 'a {}', {})
+
+  // @ts-ignore
+  expect(result._root).toBeUndefined()
+  expect(result.root.nodes).toHaveLength(1)
+
+  // @ts-ignore
+  expect(result._root).toBeDefined()
+  expect(result.root.nodes).toHaveLength(1)
+})
```
[the remark this answers is not in the document]
Comment (ai) [reply test/no-work-result.test.ts:44]: You can remove `result._root` test. We just need to call `result.root` twice and see that there is no errors.

File: lib/no-work-result.js (hunk 7 above)
Comment (ai) [reply lib/no-work-result.js:30]: There is no `LazyResult.result` so I do not know why do we have `NoWork#result` too :)

File: lib/no-work-result.js [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+      get() {
+        return self.root
+      }
+    })
```
[the remark this answers is not in the document]
Comment (ai) [reply lib/no-work-result.js:34]: It looks fine

File: lib/map-generator.js (hunk 5 above)
Comment (bogdan0083) [reply lib/map-generator.js:73]: I just noticed that with empty processor and enabled map, this:
```js
a {}
/*# sourceMappingURL=a.css.map */
b {}
/*# sourceMappingURL=b.css.map */
```

Converts to this:
```js
a {}

b {}

/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIjxubyBzb3VyY2U+Il0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBIiwiZmlsZSI6InRvLmNzcyJ9 */
```

Note the new line in the converted file before source map. It should not be there, right?

File: lib/no-work-result.js (hunk 7 above)
Comment (bogdan0083) [reply lib/no-work-result.js:26]: We don't have `NoWork#result` in public API either. It just gets returned in async promise resolve and `sync()`

File: test/no-work-result.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (bogdan0083) [reply test/no-work-result.test.ts]: Done!

File: lib/map-generator.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (bogdan0083) [reply lib/map-generator.js:187]: Done! But set line as 1 instead of 0 because line is 1-based.

File: lib/map-generator.js (hunk 5 above)
Comment (ai) [reply lib/map-generator.js:72]: Yeap.

**Comment by ai** [comment]
The last thing to fix and we are ready to merge https://github.com/postcss/postcss/pull/1659/files#r744111326

File: lib/map-generator.js (hunk 5 above)
Comment (bogdan0083) [reply lib/map-generator.js]: In the last commit I changed the regex. Now this:
```js
a {}/*# sourceMappingURL=a.css.map */
/*# sourceMappingURL=b.css.map */
b {}
/*# sourceMappingURL=c.css.map */

```
Gets converted to this:
```js
a {}
b {}
/*# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbIjxubyBzb3VyY2U+Il0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiJBQUFBIiwiZmlsZSI6InRvLmNzcyJ9 */
```

Is this correct?

Comment (ai) [reply lib/map-generator.js]: Yeap :+1:

**Pull request merged**